Transparent operation on JupyterHub

jaytmiller opened the pull request:
Modifies "show" functions to work on JupyterHub transparently, automatically implementing the scheme currently documented here:

  https://docs.bokeh.org/en/latest/docs/user_guide/output/jupyter.html#jupyterhub

 to traverse the JupyterHub proxy when the env var BOKEH_JUPYTER_EXTERNAL_URL is set by the user or hub admin. 

 This has the benefits that:

  1. Users don't need to learn anything special for Bokeh to work on JupyterHub.
  2. The same notebook will run without  modification standalone in JupyterLab or on JupyterHub

- [x ] issues: fixes #13403
- [ ] tests added / passed
- [ ] release document entry (if new feature or API change)

jaytmiller committed:
Modified "show" functions to work on JupyterHub transparently,
automatically implementing the scheme currently documented here:
  https://docs.bokeh.org/en/latest/docs/user_guide/output/jupyter.html#jupyterhub
when the env var BOKEH_JUPYTER_EXTERNAL_URL is set by the user
or hub admin.

philippjfr commented:
Awesome, thank you! I think anything that makes it work more transparently is a good thing so I think we should merge and document this. Orchestrating tests for this would be pretty involved so I don't think it should be a requirement for merging.

jaytmiller commented:
Excellent,  I will work on some docs for it tomorrow or Monday.   Thank you!

mosc9575 commented on src/bokeh/io/notebook.py:
```diff
@@ -498,6 +500,41 @@ def publish_display_data(data: dict[str, Any], metadata: dict[Any, Any] | None =
 
+def remote_jupyter_proxy_url(port: int):
+    """Callable to configure Bokeh's show method when a proxy must be
+    configured.    If port is None we're asking about the URL
```
```suggestion
    configured. If port is None we're asking about the URL
```

jaytmiller replied:
done

mosc9575 commented on src/bokeh/io/notebook.py:
```diff
@@ -498,6 +500,41 @@ def publish_display_data(data: dict[str, Any], metadata: dict[Any, Any] | None =
+    a user's environment to the external hostname of the hub, e.g. https://our-hub.edu
+
+    Args:
+       port(int):
```
```suggestion
       port (int):
```

jaytmiller replied:
done

mosc9575 commented on src/bokeh/io/notebook.py:
```diff
@@ -519,6 +556,11 @@ def show_app(app: Application, state: State, notebook_url: str | Callable[[int |
             None, it must generate the origin URL for the server.
 
+            If the environment variable BOKEH_JUPYTER_EXTERNAL_URL is set
+            to the external URL of a JupyterHub,  notebook_url is overridden
```
```suggestion
            to the external URL of a JupyterHub, notebook_url is overridden
```

mosc9575 commented on src/bokeh/io/showing.py:
```diff
@@ -115,6 +116,11 @@ def show(obj: UIElement | Application | ModifyDoc, browser: str | None = None, n
             is passed, the function is to generate the origin URL.
 
+            If the environment variable BOKEH_JUPYTER_EXTERNAL_URL is set
+            to the external URL of a JupyterHub,  notebook_url is overridden
```
```suggestion
            to the external URL of a JupyterHub, notebook_url is overridden
```

mosc9575 commented on src/bokeh/io/notebook.py:
```diff
@@ -498,6 +500,41 @@ def publish_display_data(data: dict[str, Any], metadata: dict[Any, Any] | None =
+       port(int):
+           random port generated by bokeh to avoid re-using recently closed ports
+
+       Returns: f(port: int) -> str
+           a callable which given a random port generates an external url which
+           will return through the proxy to the current notebook server.
```
```suggestion
       Returns:
           str: a callable which given a random port generates an external url which
                will return through the proxy to the current notebook server.
```

jaytmiller replied:
This code was way off so I rewrote it to capture (a) it's just a string and (b) it's a URL which will traverse the JupyterHub proxy to get back to this notebook session.

mosc9575 reviewed (commented):
I want to suggest some minor improvements in the docstrings.

bryevdv commented on src/bokeh/io/notebook.py:
```diff
@@ -498,6 +500,41 @@ def publish_display_data(data: dict[str, Any], metadata: dict[Any, Any] | None =
+           a callable which given a random port generates an external url which
+           will return through the proxy to the current notebook server.
+    """
+    base_url = os.environ['BOKEH_JUPYTER_EXTERNAL_URL']
```
Things prefixed with `BOKEH_` would normally be accessible as proper settings in [`bokeh.settings`](https://docs.bokeh.org/en/latest/docs/reference/settings.html) Should that be the case here? If not perhaps let's swap the name to `JUPYTER_BOKEH_EXTERNAL_URL` maybe? to avoid confused expectations.

jaytmiller replied:
done

bryevdv commented on src/bokeh/io/notebook.py:
```diff
@@ -498,6 +500,41 @@ def publish_display_data(data: dict[str, Any], metadata: dict[Any, Any] | None =
+    if port is None:
+        return host
+
+    service_url_path = os.environ['JUPYTERHUB_SERVICE_PREFIX']
```
This would blow up in case `os.environ['JUPYTERHUB_SERVICE_PREFIX']` does not exist? Seems like `.get` might be useful here

jaytmiller replied:
JUPYTERHUB_SERVICE_PREFIX is defined by JupyterHub so in this context it automatically exists.

bryevdv replied:
There's truly no guarantee that JupyterHub is actually running. :) The one thing about users is that they use -- anything and everything, lots functions I would have never thought anyone would ever look at someone has tried to use for some purpose, possibly intended, or often not really. Code also gets moved over the years due to refactoring or re-purposing and sometimes context around the code is lost or missed. We just really always need to be defensive, and if there is an obvious path to some error condition, put and more informative message there, and hopefully reduce the amount of traffic on the tracker or support forum.

jaytmiller committed:
Apply suggestions from code review

Co-authored-by: Moritz Schreiber <68053396+mosc9575@users.noreply.github.com>

jaytmiller committed:
Correct docstring of return value for remote_jupyter_proxy_url().
Rename env var BOKEH_JUPYTER_EXTERNAL_URL -> JUPYTER_BOKEH_EXTERNAL_URL.

bryevdv commented on src/bokeh/io/notebook.py:
```diff
@@ -498,6 +500,41 @@ def publish_display_data(data: dict[str, Any], metadata: dict[Any, Any] | None =
+           str:
+               URL capable of traversing the JupyterHub proxy to return to this notebook session.
+    """
+    base_url = os.environ['JUPYTER_BOKEH_EXTERNAL_URL']
```
Same comment here. I know you have checked for the existence of this before calling this function, but to be honest I think it might be better to not do that, and instead just have a function that encapsulates updating the `notebook_url` from the environment, that encapsulates any and all `environ` access inside it (defensively). More generally I just don't love that `show_app` is directly looking at `os.environ` itself, it will be easier to add tests for this if there is only one function that needs to mock env var settings to test. 

Alternatively,  if we want to make all these assumptions, then at least we need to make this function "private" by renaming to `_remote_jupyter_proxy_url`

jaytmiller replied:
I was debating _remote_jupyter_proxy_url myself and that sounds like the best move to me.   Coming up.

bryevdv commented:
> Orchestrating tests for this would be pretty involved so I don't think it should be a requirement for merging.

Actually running JupyterHub would be but the function for computing the notebook URL seems amendable to basic unit tests that mock `os.environ`.

bryevdv commented on src/bokeh/io/notebook.py:
```diff
@@ -498,6 +500,41 @@ def publish_display_data(data: dict[str, Any], metadata: dict[Any, Any] | None =
     from IPython.display import publish_display_data
     publish_display_data(data, metadata, transient=transient, **kwargs)
 
+def remote_jupyter_proxy_url(port: int):
```
missing return type

jaytmiller replied:
got it.

bryevdv commented on src/bokeh/io/notebook.py:
```diff
@@ -498,6 +500,41 @@ def publish_display_data(data: dict[str, Any], metadata: dict[Any, Any] | None =
     publish_display_data(data, metadata, transient=transient, **kwargs)
 
+def remote_jupyter_proxy_url(port: int):
+    """Callable to configure Bokeh's show method when a proxy must be
```
nit: all our docstrings start with a space before the sentence: 
```suggestion
    """ Callable to configure Bokeh's show method when a proxy must be
```
not the decision I'd made today but it's what we have 12 years of to be consistent with.

jaytmiller replied:
got it.

jaytmiller committed:
Renamed remote_jupyter_proxy_url to private _remote_jupyter_proxy_url.
Fixed missing space start of docstring.
Added unit test for _remote_jupyter_proxy_url based on mock os.environ.

jaytmiller committed:
Fix unit test for _remote_jupyter_proxy_url

jaytmiller committed:
Correct parameter signature for _remote_jupyter_proxy_url to include None
as an alternative to int.

bryevdv commented on src/bokeh/io/notebook.py:
```diff
@@ -498,6 +500,41 @@ def publish_display_data(data: dict[str, Any], metadata: dict[Any, Any] | None =
     from IPython.display import publish_display_data
     publish_display_data(data, metadata, transient=transient, **kwargs)
 
+def _remote_jupyter_proxy_url(port: [int | None]) -> str:
```
no brackets AFAIK
```suggestion
def _remote_jupyter_proxy_url(port: int | None) -> str:
```
also docstring is out of sync now

bryevdv commented on src/bokeh/io/notebook.py:
```diff
@@ -539,6 +581,9 @@ def show_app(app: Application, state: State, notebook_url: str | Callable[[int |
     loop = IOLoop.current()
 
+    if os.environ.get("JUPYTER_BOKEH_EXTERNAL_URL"):
+        notebook_url = _remote_jupyter_proxy_url
```
If `JUPYTER_BOKEH_EXTERNAL_URL` is set and `notebook_url` is explicitly passed in at the same time, there should probably be a warning issued. I still think a pattern like this might be preferable:
```python
notebook_url = _update_notebook_url_from_env(notebook_url)
```
so that env var handling is encapsulated in one place.

jaytmiller replied:
I'm looking at this and believe I may have found minor flaws in the type signatures of the parent functions;  in particular I was looking for the default value for notebook_url so I could determine when it is specified.  Type annotations are very new to me but it looks to me like showing.show() and notebook.show_app() alternately omit the callable or provide no default at all.  To address that,  I have defined a type alias ProxyUrl and then in the definitions of show() and show_app() default the value to "localhost:8888".   This is well outside my comfort zone so I figured I should check on this specifically.

jaytmiller replied:
Here's my type alias:

   ProxyUrl= Union[str, Callable[Union[int, None], str]]

which I use as:

   notebook_url: ProxyUrl = "localhost:8888"

so in one instance I am making up the default,  and the other adding callable (which is already documented in the docstring).

bryevdv replied:
@jaytmiller I haven't looked at the type question yet but FYI the relevant codepath goes through here:

https://github.com/bokeh/bokeh/blob/a578a98dc1c4b937cfaf89821c7f184c03b38a36/src/bokeh/io/showing.py#L152

So the default value comes from here: 

https://github.com/bokeh/bokeh/blob/a578a98dc1c4b937cfaf89821c7f184c03b38a36/src/bokeh/io/showing.py#L55

jaytmiller replied:
Thanks very much for looking at this!  This PR is addressing two entry points,  the one above defines a default but does not permit a callable,  the other one:
https://github.com/bokeh/bokeh/blob/a578a98dc1c4b937cfaf89821c7f184c03b38a36/src/bokeh/io/notebook.py#L501
permits a callable but does not define a default.   That spurred this inquiry because I wasn't sure how to test for "defined" other than to look for "not the default."

One  other thing I just noticed and didn't deal with is this:
https://github.com/bokeh/bokeh/blob/a578a98dc1c4b937cfaf89821c7f184c03b38a36/src/bokeh/io/notebook.py#L164
and the following line:
https://github.com/bokeh/bokeh/blob/a578a98dc1c4b937cfaf89821c7f184c03b38a36/src/bokeh/io/notebook.py#L165
In appears to be a related Protocol but neither mentions the callable nor defines a default.   My version of a fix would replace "str" with "ProxyUrl" and if possible add a default.   I'm wondering however if I should even think about touching a public interface.

bryevdv replied:
I would recommend this:
```python
ProxyUrlFunc: TypeAlias = Callable[[int | None], str]
```
And then in all the function signatures (I think, at a quick glance) could then be:
```python
notebook_url: str | ProxyUrlFunc
```

As for the defaults, IIRC is was an explicit decision to have defaults only in the "leaf" (user-facing) APIs and to have the params be required (no default) in the more internal APIs (the protocol `__call__` definitely does not need a default specified, it's only for `mypy` validation)

bryevdv replied:
> I'm wondering however if I should even think about touching a public interface.

Changing type annotation won't change anything at all for runtime usage of the API.

jaytmiller replied:
Your suggestion here worked out fine,  thanks!

jaytmiller replied:
Spoke too soon.   It worked fine as I deployed it,  but I wasn't testing on Python-3.9 and it looks like TypeAlias may be more recent.   What should I do here,  revert it to a simple variable assignment,  or leave it as-is?

bryevdv replied:
Have to use 
```
from typing_extensions import TypeAlias
```
for now until we are at minpy=3.11

jaytmiller replied:
Got it, thanks!

jaytmiller committed:
Update src/bokeh/io/notebook.py

Correct type annotation for port in _remote_jupyter_proxy_url.

Co-authored-by: Bryan Van de Ven <bryan@bokeh.org>

mattpap commented on src/bokeh/io/notebook.py:
```diff
@@ -13,6 +13,8 @@
 from __future__ import annotations
 
+import os
+import urllib
```
These two imports need to go under standard library imports section below.

jaytmiller replied:
Got it.

mattpap commented on src/bokeh/io/notebook.py:
```diff
@@ -498,6 +500,41 @@ def publish_display_data(data: dict[str, Any], metadata: dict[Any, Any] | None =
+        return host
+
+    service_url_path = os.environ['JUPYTERHUB_SERVICE_PREFIX']
+    proxy_url_path = 'proxy/%d' % port
```
Prefer f-strings.

jaytmiller replied:
Got it.

jaytmiller committed:
Work on transparent JupyterHub handling.

Cleaned up JupyterHub handling to addess all environment variables inside the
scope of a top level wrapper function _update_notebook_url_from_env().

Added a warning to _update_notebook_url_from_env() issued when the
show()/show_app() function's caller both defines JUPYER_BOKEH_EXTERNAL_URL and
the parameter notebook_url which is overridden and irrelevant.  Defined the
type alias ProxyUrl to describe the notebook_url parameter being passed to
show()/show_app() functions which is either a string or callable and now
defaults to "localhost:8888".

Cleaned up Python imports.  Switched from % formatting to f-string.

jaytmiller committed:
Removed unused import os from bokeh.io.showing

jaytmiller committed:
Address isort issues in bokeh.io .notebook and .showing

jaytmiller committed:
Converted ProxyUrl to ProxyUrlFunc to improve notebook_url interfaces.

jaytmiller committed:
Added unit tests for _update_notebook_url_from_env

jaytmiller committed:
Add test for _remote_jupyter_proxy_url(None)

jaytmiller committed:
Removed _update_notebook_url_from_env call from showing.py so it doesn't
get called twice.

jaytmiller committed:
Updated the docstring for _update_notebook_url_from_env

jaytmiller committed:
Updated docs regarding JUPYTER_BOKEH_EXTERNAL_URL.

mattpap commented:
> Robustify asset resolution in scale_bar_image.py [31b5f5b]
> Deployment updates for release 3.3.0.dev3 [c4c12ae]

These commits shouldn't be on the list. I'm guessing bad merge or rebase? Let's remove them for the sake of clarity.

jaytmiller commented:
I've been having trouble with the various tool chains so I think I did rebase.  Rebasing again and drop'ing the commits seemed to work,  but it looks like I have to force push to use that method.  Is there a better way or is that OK?

bryevdv commented:
This is a pretty small PR I think a force-push would be fine

jaytmiller committed:
Switched import of TypeAlias to typing_extensions

bryevdv commented on src/bokeh/io/showing.py:
```diff
@@ -52,7 +52,7 @@
 
 def show(obj: UIElement | Application | ModifyDoc, browser: str | None = None, new: BrowserTarget = "tab",
-        notebook_handle: bool = False, notebook_url: str = "localhost:8888", **kwargs: Any) -> CommsHandle | None:
+        notebook_handle: bool = False, notebook_url: [str | ProxyUrlFunc] = "localhost:8888", **kwargs: Any) -> CommsHandle | None:
```
No braces
```suggestion
        notebook_handle: bool = False, notebook_url: str | ProxyUrlFunc = "localhost:8888", **kwargs: Any) -> CommsHandle | None:
```

bryevdv commented on src/bokeh/io/notebook.py:
```diff
@@ -498,7 +501,10 @@ def publish_display_data(data: dict[str, Any], metadata: dict[Any, Any] | None =
+
+ProxyUrlFunc: TypeAlias = Callable[[int | None], str]
+
+def show_app(app: Application, state: State, notebook_url: [str | ProxyUrlFunc] = "localhost:8888", port: int = 0, **kw: Any) -> None:
```
No braces 
```suggestion
def show_app(app: Application, state: State, notebook_url: str | ProxyUrlFunc = "localhost:8888", port: int = 0, **kw: Any) -> None:
```

bryevdv replied:
Also we have a fairly long line length limit in order to accommodate some things, but this line is now long enough that this might be prefered
```python
def show_app(
    app: Application, 
    state: State, 
    notebook_url: str | ProxyUrlFunc = "localhost:8888", 
    port: int = 0, 
    **kw: Any
) -> None:
```

bryevdv commented on src/bokeh/io/notebook.py:
```diff
@@ -640,6 +653,69 @@ def _server_url(url: str, port: int | None) -> str:
+    return full_url
+
+
+def _update_notebook_url_from_env(notebook_url: [str | ProxyUrlFunc]) -> [str | ProxyUrlFunc]:
```
no braces
```suggestion
def _update_notebook_url_from_env(notebook_url: str | ProxyUrlFunc) -> str | ProxyUrlFunc:
```

bryevdv commented on src/bokeh/io/notebook.py:
```diff
@@ -640,6 +653,69 @@ def _server_url(url: str, port: int | None) -> str:
+          when ``JUPYER_BOKEH_EXTERNAL_URL`` is set.
+    """
+    if os.environ.get("JUPYTER_BOKEH_EXTERNAL_URL"):
+        if notebook_url != "localhost:8888":
```
If we want to check against a default value that is not None, then we will need to define something like
```
DEFAULT_JUPYTER_URL = "localhost:8888"
```
somewhere, and then use that here, and also wherever the defaults are specified, in order to avoid duplication in multiple places (and the possibility that in the future one gets updated but not another)

bryevdv reviewed (commented):
getting close, a few more comments

bryevdv commented on docs/bokeh/source/docs/user_guide/output/jupyter.rst:
```diff
@@ -157,7 +193,7 @@ instructions above, then continue with the following steps:
         # so return the public hostname.
         if port is None:
-            return host
+           return host
```
This seems amiss, we typically use 4-space indents

bryevdv commented:
@philippjfr I don't have any further outstanding comments, did you want have a look-over?

jaytmiller committed:
Update src/bokeh/io/showing.py

Co-authored-by: Bryan Van de Ven <bryan@bokeh.org>

jaytmiller committed:
Update src/bokeh/io/notebook.py

Co-authored-by: Bryan Van de Ven <bryan@bokeh.org>

jaytmiller committed:
Update src/bokeh/io/notebook.py

Co-authored-by: Bryan Van de Ven <bryan@bokeh.org>

jaytmiller committed:
Convert inline "localhost:8888" defaults to DEFAULT_JUPYTER_URL
Address excessive line length in show_app def.

jaytmiller committed:
Fixed bad indent in jupyter.rst

bryevdv commented:
@jaytmiller Our build job runs `bokeh info` as a sanity check and annoyingly this new `typing_extensions` import is in that codepath, so our build env file needs to add that package now:

https://github.com/bokeh/bokeh/blob/branch-3.3/conda/environment-build.yml

Honestly surprised we have no run into this before now, actually. 

It's already a runtime dependency in the other conda env files, you can refer to those (need a min ver IIRC) or let me know if any questions.

mattpap commented:
We import `typing_extensions` behind `if TYPE_CHECKING:` for that reason. Given that annotations are lazily computed, then it means that `TypeAlias` can be used outside of `if TYPE_CHECKING:`, for not making the code unbearable to read.

mattpap commented on src/bokeh/io/notebook.py:
```diff
@@ -35,6 +37,7 @@
 from uuid import uuid4
 
 ## External imports
+from typing_extensions import TypeAlias
 if TYPE_CHECKING:
     from ipykernel.comm import Comm
```
```suggestion
if TYPE_CHECKING:
    from ipykernel.comm import Comm
    from typing_extensions import TypeAlias
```

bryevdv replied:
yup this is the correct solution, had a brain freeze earlier I guess!

jaytmiller committed:
Update src/bokeh/io/notebook.py

Co-authored-by: Mateusz Paprocki <mattpap@gmail.com>

jaytmiller commented:
Any suggestions on the failed Bokeh-CI?   My guess is that "|" in type annotations isn't fully supported until Python-3.10,  so my fix for 3.9 would be:

    ProxyUrlFunc: TypeAlias = Callable[[int| None], str]

becomes:

    ProxyUrlFunc: TypeAlias = Callable[[Union[int, None]], str]

Sound OK?

jaytmiller committed:
Updated def of ProxyUrlFunc type alias to avoid using | for Python-3.9

bryevdv commented:
@jaytmiller yes, that's the right solution. Even with `__future__` in 3.9 the use of `|` is restricted to places that can be lazily evaluated (not here).

jaytmiller commented:
OK, ProxyUrlFunc updated.   At this stage I believe I've addressed all the issues mentioned so far,  please let me know if there are any outstanding concerns.  Thank you for your help and patience!

jaytmiller committed:
Added ProxyUrlFunc back to showing.py

jaytmiller commented:
Sorry,  spoke too soon,  accidentally removed a required ProxyUrlFunc import in showing.py.  Pushed

jaytmiller committed:
Addresses isort issues

bryevdv commented:
Thanks for your perserverance @jaytmiller !

@mattpap I am going to go ahead and merge this to `branch-3.3` and cut a new rc

jaytmiller commented:
Thanks guys for all your help getting this done!
